File: backend/src/main/java/com/maksimenko/Clicker/model/StockCostGenerator.java
package com.maksimenko.Clicker.model;

import com.maksimenko.Clicker.utils.StockUtils;

public class StockCostGenerator {

    private double currentLength = 0d;
    private double maxLength = 0d;
    private double maxRadius = 0d;
    private boolean activeCalculation = false;
    private double value = 0d;

    private boolean positiveRise = true;

    public double generateCost(double actualCost){
        if (activeCalculation) {
            currentLength++;

            double changedValue = maxRadius * Math.sin(Math.PI * 0.5 * (currentLength / maxLength));

            if (positiveRise) {
                value = actualCost + changedValue;
            } else {
                value = actualCost - changedValue;
            }

            if (currentLength > maxLength) {
                activeCalculation = false;
            }
        } else {
            currentLength = 0d;

            positiveRise = !positiveRise;

            maxRadius = Math.random() * (actualCost * 0.2);
            maxLength = Math.random() * 20;

            activeCalculation = true;
        }

        return StockUtils.round2(value);
    }
}

File: frontend/src/utils/StockCostGenerator.js
import { Utils } from './Utils'

export class StockCostGenerator {
    constructor() {
        this.currentLength = 0;
        this.maxLength = 0;
        this.maxRadius = 0;
        this.activeCalculation = false;
        this.value = 0;
        this.positiveRise = true;
    }

    generateCost(actualCost, symbol) {
        if (this.activeCalculation) {
            this.currentLength++;

            let changedValue = this.maxRadius * Math.sin(Math.PI * (this.currentLength / this.maxLength));

            if (this.positiveRise) {
                this.value = actualCost + changedValue;
            } else {
                this.value = actualCost - changedValue;
            }

            if (this.currentLength > this.maxLength) {
                this.activeCalculation = false;
            }
        } else {
            this.currentLength = 0;

            this.positiveRise = !this.positiveRise;

            this.maxRadius = Math.random() * (actualCost * 0.2);
            this.maxLength = Math.random() * 40;

            this.activeCalculation = true;
        }

        return Utils.round2(this.value);
    }
}